pkg/tbtc/registry.go

```diff
@@ -0,0 +1,195 @@
+package tbtc
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"encoding/hex"
+	"fmt"
+	"github.com/keep-network/keep-common/pkg/persistence"
+	"sync"
+)
+
+// walletRegistry is the component that holds the data of the wallets managed
+// by the given node. All functions of the registry are safe for concurrent use.
+type walletRegistry struct {
+	// mutex is a single struct-wide lock that ensures all functions
+	// of the registry are thread-safe.
+	mutex sync.Mutex
+
+	// walletCache is a cache of maintained wallets. The cache's key is the
+	// uncompressed public key of the given wallet. The cache's value is
+	// a slice of the wallet signers controlled by this node.
+	walletCache map[string][]*signer
+
+	// walletStorage is the handle to the wallet storage responsible for
+	// wallet persistence.
+	walletStorage *walletStorage
+}
+
+// newWalletRegistry creates a new instance of the walletRegistry.
+func newWalletRegistry(persistence persistence.Handle) *walletRegistry {
+	walletStorage := newWalletStorage(persistence)
+
+	// Pre-populate the wallet cache using the wallet storage.
+	walletCache := walletStorage.loadSigners()
+	if len(walletCache) > 0 {
+		for walletStorageKey, signers := range walletCache {
+			logger.Infof(
+				"wallet signing group [0x%v] loaded from storage "+
+					"with [%v] members",
+				walletStorageKey,
+				len(signers),
+			)
+		}
+	} else {
+		logger.Infof("no wallet signing groups found in the storage")
+	}
+
+	return &walletRegistry{
+		walletCache:   walletCache,
+		walletStorage: walletStorage,
+	}
+}
+
+// registerSigner registers the given signer using in the walletRegistry.
+func (wr *walletRegistry) registerSigner(signer *signer) error {
+	wr.mutex.Lock()
+	defer wr.mutex.Unlock()
+
+	err := wr.walletStorage.saveSigner(signer)
+	if err != nil {
+		return fmt.Errorf("cannot save signer in the storage: [%w]", err)
+	}
+
+	walletStorageKey := getWalletStorageKey(signer.wallet.publicKey)
+
+	wr.walletCache[walletStorageKey] = append(
+		wr.walletCache[walletStorageKey],
+		signer,
+	)
+
+	return nil
+}
+
+// getSigners gets all signers for the given wallet held by the walletRegistry.
+func (wr *walletRegistry) getSigners(
+	walletPublicKey *ecdsa.PublicKey,
+) []*signer {
+	wr.mutex.Lock()
+	defer wr.mutex.Unlock()
+
+	return wr.walletCache[getWalletStorageKey(walletPublicKey)]
+}
+
+// walletStorage is the component that persists data of the wallets managed
+// by the given node using the underlying persistence layer. It should be
+// used directly only by the walletRegistry.
+type walletStorage struct {
+	// persistence is the handle to the underlying persistence layer.
+	persistence persistence.Handle
+}
+
+// newWalletStorage creates a new instance of the walletStorage.
+func newWalletStorage(persistence persistence.Handle) *walletStorage {
+	return &walletStorage{persistence}
+}
+
+// saveSigner saves the given signer using the underlying persistence layer
+// of the walletStorage.
```

Comment:
Nitpicking but I think it would be worth noting in the documentation this function is not adding signer to the cache and should not be called from the outside of the struct.

Reply:
See 717040f1